### ISSUE TITLE
Energy measurement results in segfault

### ISSUE THREAD

**Issue by d4k0** [RRZE-HPC/pylikwid#9]
Hi,

I wanted to measure the energy using the code in the energy paragraph of the pylikwid GitHub page, but somehow it always results in a crash and I don't know why. Here is a small sample code:

```
import pylikwid

if __name__ == '__main__':
    list = []
    cpuid = 0
    pinfo = pylikwid.getpowerinfo()
    domainid = pinfo.get("domains").get("PKG").get("ID")

    e_start = pylikwid.startpower(cpuid, domainid)

    # for i in range(0, 1000000):
    #     list.append(i)

    e_stop = pylikwid.stoppower(cpuid, domainid)

    e = pylikwid.getpower(e_start, e_stop, domainid)

    pylikwid.putpowerinfo() # not sure if this is needed

    print(e)
```

Running this in PyCharm or on the terminal (both with `sudo`) gives me this (the loop is commented out because it "hides" some of the errors):

```
/home/user/PycharmProjects/Tests/venv/bin/python2.7 /home/user/PycharmProjects/Tests/EnergyTest2.py
ERROR - [./src/includes/power.h:97] MSR read operation failed - No such file or directory 
ERROR - [./src/includes/power.h:124] MSR read operation failed - No such file or directory 

Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)
```

The crash seems to occur when using the `getpower` function. In the `syslog` I found the following error messages after running the program:

```
Jun 14 08:48:03 user-HP-ProBook-450-G0 accessD: AccessDaemon runs with UID 0, eUID 0
Jun 14 08:48:03 user-HP-ProBook-450-G0 accessD: Failed to read data from register 0x613 on core 0
Jun 14 08:48:03 user-HP-ProBook-450-G0 accessD: Failed to read data from register 0x63b on core 0
Jun 14 08:48:03 user-HP-ProBook-450-G0 accessD: Failed to read data from register 0x619 on core 0
Jun 14 08:48:03 user-HP-ProBook-450-G0 accessD: Failed to read data from register 0x620 on core 0
Jun 14 08:48:03 user-HP-ProBook-450-G0 kernel: [17662.614324] python2.7[14371]: segfault at 7fc895be1ed8 ip 00007fc894f5f4df sp 00007ffca6a30f20 error 4 in liblikwid.so.4.3[7fc894da7000+253000]
Jun 14 08:48:03 user-HP-ProBook-450-G0 accessD: ERROR - [accessDaemon.c:1737] zero read
Jun 14 08:48:03 user-HP-ProBook-450-G0 kernel: [17662.937725] likwid-accessD[14386]: segfault at 19 ip 00007fb02d85098d sp 00007ffcd85a8150 error 4 in libc-2.27.so[7fb02d7b9000+1e7000]
```

I then saw that the mode (direct or daemon) can be set in the code and I tried the following:

```
import pylikwid

if __name__ == '__main__':
    pylikwid.hpmmode(0)
    pylikwid.hpminit()
    pylikwid.hpmaddthread(0)
    pylikwid.hpmaddthread(1)
    pylikwid.hpmaddthread(2)
    pylikwid.hpmaddthread(3)

    list = []
    cpuid = 0
    pinfo = pylikwid.getpowerinfo()
    domainid = pinfo.get("domains").get("PKG").get("ID")

    e_start = pylikwid.startpower(cpuid, domainid)

    # for i in range(0, 1000000):
    #     list.append(i)

    e_stop = pylikwid.stoppower(cpuid, domainid)

    e = pylikwid.getpower(e_start, e_stop, domainid)

    pylikwid.hpmfinalize()
    pylikwid.putpowerinfo()

    print(e)
```

Running it gives me the following:

```
/home/user/PycharmProjects/Tests/venv/bin/python2.7 /home/user/PycharmProjects/Tests/EnergyTest.py
/dev/cpu/0/msr
/dev/cpu/1/msr
/dev/cpu/2/msr
/dev/cpu/3/msr

Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)
```

It seems to load the msr files now, but it still gives me a segfault. The syslog now shows this:

`Jun 14 08:47:06 user-HP-ProBook-450-G0 kernel: [17605.663362] python2.7[14358]: segfault at 7f5969fd4458 ip 00007f59693524df sp 00007ffd6905fcf0 error 4 in liblikwid.so.4.3[7f596919a000+253000]`

**Do I have to to initialize something in the code before? Or could this be a user right problem?** The strange thing is that `likwid-powermeter` works:

```
$ sudo likwid-powermeter --------------------------------------------------------------------------------
CPU name:	Intel(R) Core(TM) i5-3230M CPU @ 2.60GHz
CPU type:	Intel Core IvyBridge processor
CPU clock:	2.59 GHz
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
Runtime: 2.00073 s
Measure for socket 0 on CPU 0
Domain PKG:
Energy consumed: 8.7892 Joules
Power consumed: 4.39299 Watt
Domain PP0:
Energy consumed: 3.50433 Joules
Power consumed: 1.75153 Watt
Domain PP1:
Energy consumed: 0.596512 Joules
Power consumed: 0.298147 Watt
```

On another laptop with a Sandy Bridge processor I get the same results. Here's some more information about my system:

```
OS: Ubuntu 18.04 (freshly installed)
gcc: 7.3.0
LIKWID: 4.3.1 (installed using apt)
Python: 2.7.15
```

**Comment by TomTheBear** [comment]
It seems there are some more problems introduced with Ubuntu 18.04, not just the changed installation paths.
I have to dig a little to see where the problem is.
Thanks for reporting.

**Comment by d4k0** [comment]
As a temporary workaround I now use `likwid-powermeter` as a wrapper and parse the output to get the energy value. Nevertheless, it is pretty strange that this works and the Python API doesn't.

What I also noticed is that the Ubuntu crash reporter sometimes shows up when I do this for the first time (?) during a session (I have to test this a bit more), but it still works and measures the energy:

![Crash1](https://i.imgur.com/uoHxTSq.png)

![Crash2](https://i.imgur.com/CnxkQmP.png)

I think I also got this a few times when I used the Python API, but as I reported it resulted in a segfault.

**One more note:** I talked to a person who is using the C API and he also gets a segfault when trying to get the energy values. He is using the computers of the chair which use openSUSE with (if I remember correctly) kernel 4.12.x.
